numcosmo/math/ncm_stats_dist_nd.c

```diff
@@ -138,15 +253,347 @@ ncm_stats_dist_nd_class_init (NcmStatsDistNdClass *klass)
+  ncm_matrix_clear (&self->log_cov);
+  ncm_matrix_clear (&self->sample_matrix);
+  ncm_vector_clear (&self->v);
+
+  self->d          = dim;
+  self->sample     = ncm_stats_vec_new (dim, NCM_STATS_VEC_COV, TRUE);
+  self->cov_decomp = ncm_matrix_new (dim, dim);
+  self->log_cov    = ncm_matrix_new (dim, dim);
+  self->v          = ncm_vector_new (dim);
+}
+
+static void
+_ncm_stats_dist_nd_prepare_kernel_args (NcmStatsDistNd *dnd, NcmStatsVec *sample)
+{
+  NcmStatsDistNdPrivate * const self = dnd->priv;
+  gint i, ret;
+
+  self->n = ncm_stats_vec_nitens (sample);
+  g_assert_cmpuint (self->n, >, 1);
+
+  ncm_matrix_memcpy (self->cov_decomp, ncm_stats_vec_peek_cov_matrix (sample, 0));
+  if (ncm_matrix_cholesky_decomp (self->cov_decomp, 'U') != 0)
+  {
+    if (ncm_matrix_nearPD (self->cov_decomp, 'U', TRUE, self->nearPD_maxiter) != 0)
+    {
+      ncm_matrix_set_zero (self->cov_decomp);
+      for (i = 0; i < self->d; i++)
+      {
+        ncm_matrix_set (self->cov_decomp, i, i, ncm_stats_vec_get_cov (sample, i, i));
+      }
+      g_assert_cmpint (ncm_matrix_cholesky_decomp (self->cov_decomp, 'U'), ==, 0);
+    }
+  }
+
+  self->href          = self->over_smooth * ncm_stats_dist_nd_get_rot_bandwidth (dnd, self->d, self->n);
+  self->href2         = self->href * self->href;
+  self->kernel_lnnorm = ncm_stats_dist_nd_get_kernel_lnnorm (dnd, self->cov_decomp, self->d, self->n, self->href);
+
+  if ((self->sample_matrix == NULL) || (self->n != ncm_matrix_nrows (self->sample_matrix)) || (self->d != ncm_matrix_ncols (self->sample_matrix)))
+  {
+    ncm_matrix_clear (&self->sample_matrix);
+    self->sample_matrix = ncm_matrix_new (self->n, self->d);
+
+    g_ptr_array_set_size (self->invUsample, 0);
+    for (i = 0; i < self->n; i++)
+    {
+      NcmVector *row_i = ncm_matrix_get_row (self->sample_matrix, i);
+      g_ptr_array_add (self->invUsample, row_i);
+    }
+  }
+
+  for (i = 0; i < self->n; i++)
+    ncm_matrix_set_row (self->sample_matrix, i, ncm_stats_vec_peek_row (self->sample, i));
+
+  ret = gsl_blas_dtrsm (CblasRight, CblasUpper, CblasNoTrans, CblasNonUnit, 1.0, ncm_matrix_gsl (self->cov_decomp), ncm_matrix_gsl (self->sample_matrix));
+  NCM_TEST_GSL_RESULT ("_ncm_stats_dist_nd_prepare_kernel_args", ret);
+}
+
+static void
+_ncm_stats_dist_nd_prepare (NcmStatsDistNd *dnd)
+{
+  NcmStatsDistNdClass *dnd_class = NCM_STATS_DIST_ND_GET_CLASS (dnd);
+  NcmStatsDistNdPrivate * const self = dnd->priv;
+
+  dnd_class->prepare_kernel_args (dnd, self->sample);
+
+  self->min_m2lnp = 0.0;
+  self->max_m2lnp = 0.0;
+
+  if (self->weights == NULL)
+  {
+    self->weights = ncm_vector_new (self->n);
+  }
+  else if (self->n != ncm_vector_len (self->weights))
+  {
+    ncm_vector_clear (&self->weights);
+    self->weights = ncm_vector_new (self->n);
+  }
+
+  ncm_vector_set_all (self->weights, 1.0 / (1.0 * self->n));
+}
+
+void LowRankQP (gint *n, gint *m, gint *p, gint *method, gint *verbose, gint *niter,
```

Comment:
I have a question here. For the LowRankQP and the nnls_c right under this code. Here you are just defining a prototype of the function right? But I didn't find the part of the code where you truly describes these functions. I know that there are the nnls.c file and a lowrankQP.c file inside the NumCosmo library, but you didn't include these files in the beginning of the code. And if we were to use only these files that are ready, why do we need to create a prototype? Am I missing something?